src/Language/PureScript/AST/Binders.hs

```diff
@@ -12,80 +12,97 @@ import Language.PureScript.Comments
 import Language.PureScript.Types
 
 -- |
--- Data type for binders
+-- Data type for binders.
 --
-data Binder
+-- The @a@ parameter allows for types to be annotated  with extra data - source
+-- position info, for example, and the @b@ parameter is for value-level
+-- annotations. Two separate annotation types are supported as not every
+-- annotation for a type would be suitable for a value, and vice-versa.
+--
+data Binder a b
   -- |
   -- Wildcard binder
   --
-  = NullBinder
+  = NullBinder b
   -- |
   -- A binder which matches a literal
   --
-  | LiteralBinder (Literal Binder)
+  | LiteralBinder (Literal (Binder a b)) b
   -- |
   -- A binder which binds an identifier
   --
-  | VarBinder Ident
+  | VarBinder Ident b
   -- |
   -- A binder which matches a data constructor
   --
-  | ConstructorBinder (Qualified (ProperName 'ConstructorName)) [Binder]
+  | ConstructorBinder (Qualified (ProperName 'ConstructorName)) [Binder a b] b
   -- |
   -- A operator alias binder. During the rebracketing phase of desugaring,
   -- this data constructor will be removed.
   --
-  | OpBinder (Qualified (OpName 'ValueOpName))
+  | OpBinder (Qualified (OpName 'ValueOpName)) b
   -- |
   -- Binary operator application. During the rebracketing phase of desugaring,
   -- this data constructor will be removed.
   --
-  | BinaryNoParensBinder Binder Binder Binder
+  | BinaryNoParensBinder (Binder a b) (Binder a b) (Binder a b) b
   -- |
   -- Explicit parentheses. During the rebracketing phase of desugaring, this
   -- data constructor will be removed.
   --
   -- Note: although it seems this constructor is not used, it _is_ useful,
   -- since it prevents certain traversals from matching.
   --
-  | ParensInBinder Binder
+  | ParensInBinder (Binder a b) b
   -- |
   -- A binder which binds its input to an identifier
   --
-  | NamedBinder Ident Binder
+  | NamedBinder Ident (Binder a b) b
   -- |
   -- A binder with source position information
   --
-  | PositionedBinder SourceSpan [Comment] Binder
+  | PositionedBinder SourceSpan [Comment] (Binder a b) b
   -- |
   -- A binder with a type annotation
   --
-  | TypedBinder Type Binder
-  deriving (Show, Eq)
+  | TypedBinder (Type a) (Binder a b) b
+  deriving (Show, Functor)
```

Comment:
Maybe you can do it without TH if you derive `Traversable X` and `Traversable (Flip X)` 😄